dq-game/src/server/controller.py

```diff
@@ -1,26 +1,100 @@
-import sys
 import os
-from utils.socket_connection import ServerSocketConnection
+import pickle
+import sys
+
 from events.event_handler import EventHandler
+from utils.socket_connection import ServerSocketConnection
 
 
 class Controller(EventHandler):
-    def __init__(self, no_of_players):
+    def __init__(self, socket, no_of_players):
         """
         Creates a controller
         """
+        self.socket = socket
+        self.socket.on("GAME_READY_TO_START", self.start)
         self.no_of_players = no_of_players
-        self.socket = ServerSocketConnection(8000)
-        self.socket.on("game_ready_to_start", self.start_game)
+        self.players = []
+
+    def start(self):
+        self.trigger("START_GAME")
+
+    def request_theme_choices(self, theme_list):
+
+        self.socket.send_to_all(theme_list, "data-theme-list")
+        self.socket.send_to_all("CHOOSE_THEME", "event")
+        self.trigger("THEMES_REQUESTED")
+
+    def get_theme_choices(self):
+        result = []
+        for message in self.socket.inbuffer:
+            if message.content_type == "THEME_CHOICE":
+                result.append(message.data)
+        print(result)
+        return result
 
-    def listen(self):
+    def decide_themes(self, theme_choices, result_size):
+        count_dict = {}
+        result = []
+        for theme_choice in theme_choices:
+            for theme in theme_choice:
+                if theme not in count_dict:
+                    count_dict[theme] = 1
+                else:
+                    count_dict[theme] += 1
+        for theme in count_dict.keys():
+            idx = 0
+            if len(result) == 0:
+                result.append(theme)
+            else:
+                while idx < len(result) and count_dict[theme] < count_dict[result[idx]]:
+                    idx += 1
+                result.insert(idx, theme)
+        return result[0:3]
```

Comment:
O(n^2)

You can do `THEME_NUMBER_OPTIONS=3` get Math.max(count_dict) => 3 O(n)